### ISSUE TITLE
Routes with optional parameters captures wrong patterns

### ISSUE THREAD

**Issue by lndl** [sinatra/sinatra#1023]
Hi,

I've the following definition for a simple route pattern using an optional parameter to retrieve the format (exactly as figure in the readme example):

```ruby
get '/some_path.?:format?' do
  "Format is #{params[:format]}\n"
end
```

It works great for a canonical case, like:

```
curl http://localhost:4567/some_path.html

Format is html
```

But actually the problem is that it's take all the characters next to the fixed part of the route, even if I'm trying to go to a non-existent route. By example:

```
curl -v http://localhost:4567/some_path_which_does_not_exist
GET /some_path_which_does_not_exist HTTP/1.1
User-Agent: curl/7.38.0
Host: localhost:4567
Accept: */*

HTTP/1.1 200 OK
Content-Type: text/html;charset=utf-8
Content-Length: 32
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Connection: keep-alive
Server thin is not blacklisted
Server: thin

Format is _which_does_not_exist
```

It is a bug or I'm missing something? I know that I could rewrite this with a regex, but I find this issue somewhat confusing.

By the way, i'm using Sinatra from master branch and Ruby 2.0.0p645

Thanks!

**Comment by JonRowe** [comment]
you need to escape the dot, or make it non optional, the `.?` is interpreted as `.` 0 or 1 times.

**Comment by lndl** [comment]
Make the dot mandatory is not an option because my application routes manages a default format when  it is not specified.
Maybe there is no way to express what I want using a route in string fashion, because what I'd need is something like this:
```ruby
get '/some_path(.:format)?' # an optional format (extension), but when is present it should have a dot
```
I thought that the example shown in the readme was behaving like that (I got carried away), but reading more precisely the regex-like-string I realized that that expresion was semantically corrent.
My question is: have I got another approach mantaining the string format, or forcibly I have to use a real regex?

Thanks again!

**Comment by JonRowe** [comment]
> My question is: have I got another approach maintaining the string format, or forcibly I have to use a real regex?

I don't think so, you could double up on the string routes but using a regex is probably preferable.

**Comment by mhutter** [comment]
I think this is actually a Bug in the README. The `.` needs to be escaped. Even though the route in the example and the one you described are written as plain strings, they're still parsed like a regex (and thus `.` is not treated as a literal dot but as a regex).

If you wrote

```ruby
get '/some_path\.?:format'
```

then `/some_path_which_does_not_exist` won't match anymore. This is also how you usually go about optional format parameters.

To match the Format INCLUDING the dot, you have two options:

```ruby
# This will NOT match '/some_path_which_does_not_exist'!
get %r{/some_path(\..*)} do |format|
    "Format is #{format}\n" # => "Format is .html\n"
end

# This WILL match '/some_path_which_does_not_exist'!
get '/some_path:format?' do
  "Format is #{params[:format]}\n" # => "Format is .html\n"
end
```

Hope this helps!

**Comment by kytrinyx** [comment]
Thanks @mhutter. Including the escaped period in the README example would be a good improvement.

**Comment by JonRowe** [comment]
I tried this but it doesn't work for me, `get '/posts\.:format' { ... }` raises an error in the regular expression compiler

**Comment by mhutter** [comment]
I was actually wrong: It's not that `.` matches everything, but because it's optional, routes without a dot will also match.

Now, for most cases, `'/posts.?:format?'` will do just fine. If you really want to make sure there is a dot (or want to include the dot in `format`, just use a regex. (your route will be compiled into a regex anyway)

**Comment by lndl** [comment]
Thanks to all for your feedback. Finally, I opted to use a custom route matcher (I could do it with a simple regex as well).
In my very humble opinion, I don't see much sense to use a optional wildcard (the ? symbol) inside a string-based route, at least for that specific case in which I had to capture the route format (perhaps I'm a little biased). Maybe a better example in the README for that feature would help.
